File: android/app/src/main/kotlin/com/TheVoteApp/data/network/LiveTallyClient.kt
package com.TheVoteApp.data.network

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject

/**
 * Connects to the backend `/ws/tally` WebSocket endpoint and emits
 * real-time [TallyUpdate] objects for subscribed policies via a [SharedFlow].
 *
 * Usage:
 * ```kotlin
 * val client = LiveTallyClient(okHttpClient, "wss://api.thevoteapp.org/ws/tally")
 * client.subscribe("pol-001")
 * lifecycleScope.launch {
 *     client.updates.collect { update -> /* update UI */ }
 * }
 * client.connect()
 * ```
 */
class LiveTallyClient(
    private val okHttpClient: OkHttpClient,
    private val wsUrl: String,
) {

    @Serializable
    data class PolicyTallyDTO(
        val policyId: String,
        val support: Int,
        val oppose: Int,
        val abstain: Int,
        val total: Int,
        val lastBlock: Int,
        val finalized: Boolean,
    )

    @Serializable
    data class TallyUpdate(
        val type: String,
        val policyId: String,
        val tally: PolicyTallyDTO,
        val blockNumber: Int? = null,
    )

    enum class State { DISCONNECTED, CONNECTING, CONNECTED }

    // Public state
    var state: State = State.DISCONNECTED
        private set

    private val _updates = MutableSharedFlow<TallyUpdate>(extraBufferCapacity = 64)
    val updates: SharedFlow<TallyUpdate> = _updates.asSharedFlow()

    private var webSocket: WebSocket? = null
    private val subscribedPolicies = mutableSetOf<String>()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val json = Json { ignoreUnknownKeys = true }

    // MARK: - Connection

    fun connect() {
        if (state != State.DISCONNECTED) return
        state = State.CONNECTING
        val request = Request.Builder().url(wsUrl).build()
        webSocket = okHttpClient.newWebSocket(request, listener)
    }

    fun disconnect() {
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        state = State.DISCONNECTED
    }

    fun subscribe(policyId: String) {
        subscribedPolicies.add(policyId)
        if (state == State.CONNECTED) {
            send(mapOf("type" to "subscribe", "policyId" to policyId))
        }
    }

    fun unsubscribe(policyId: String) {
        subscribedPolicies.remove(policyId)
        if (state == State.CONNECTED) {
            send(mapOf("type" to "unsubscribe", "policyId" to policyId))
        }
    }

    fun close() {
        disconnect()
        scope.cancel()
    }

    // MARK: - Private

    private fun send(map: Map<String, String>) {
        webSocket?.send(JSONObject(map).toString())
    }

    private val listener = object : WebSocketListener() {
        override fun onOpen(ws: WebSocket, response: Response) {
            state = State.CONNECTED
            // Re-subscribe to all tracked policies on connect/reconnect
            subscribedPolicies.forEach { id ->
                send(mapOf("type" to "subscribe", "policyId" to id))
            }
        }

        override fun onMessage(ws: WebSocket, text: String) {
            try {
                val update = json.decodeFromString<TallyUpdate>(text)
                if (update.type == "tally" || update.type == "snapshot") {
                    scope.launch { _updates.emit(update) }
                }
            } catch (_: Exception) { /* ignore malformed messages */ }
        }

        override fun onClosing(ws: WebSocket, code: Int, reason: String) {
            state = State.DISCONNECTED
        }

        override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
            state = State.DISCONNECTED
        }
    }
}

File: ios/VoteGlobal/Data/Network/LiveTallyClient.swift
import Foundation
import Combine

/// Connects to the backend `/ws/tally` WebSocket endpoint and publishes
/// real-time `PolicyTally` updates for subscribed policies.
///
/// Usage:
/// ```swift
/// let client = LiveTallyClient(url: URL(string: "wss://api.thevoteapp.org/ws/tally")!)
/// client.subscribe(to: "pol-001")
/// client.tallyPublisher
///     .receive(on: DispatchQueue.main)
///     .sink { update in … }
///     .store(in: &cancellables)
/// await client.connect()
/// ```
@MainActor
public final class LiveTallyClient: NSObject, ObservableObject {

    // MARK: - Types

    public struct TallyUpdate: Decodable {
        public let type: String
        public let policyId: String
        public let tally: PolicyTallyDTO
        public let blockNumber: Int?
    }

    public struct PolicyTallyDTO: Decodable {
        public let policyId: String
        public let support: Int
        public let oppose: Int
        public let abstain: Int
        public let total: Int
        public let lastBlock: Int
        public let finalized: Bool
    }

    // MARK: - Public

    @Published public private(set) var connectionState: ConnectionState = .disconnected

    public enum ConnectionState { case disconnected, connecting, connected }

    public let tallySubject = PassthroughSubject<TallyUpdate, Never>()

    // MARK: - Private

    private let url: URL
    private var webSocketTask: URLSessionWebSocketTask?
    private var subscribedPolicies: Set<String> = []
    private let session = URLSession(configuration: .default)
    private let decoder = JSONDecoder()

    public init(url: URL) {
        self.url = url
    }

    // MARK: - Connection

    public func connect() async {
        guard connectionState == .disconnected else { return }
        connectionState = .connecting
        let task = session.webSocketTask(with: url)
        webSocketTask = task
        task.resume()
        connectionState = .connected

        // Re-subscribe to any already-tracked policies
        for policyId in subscribedPolicies {
            await send(["type": "subscribe", "policyId": policyId])
        }

        await receiveLoop()
    }

    public func disconnect() {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        connectionState = .disconnected
    }

    public func subscribe(to policyId: String) {
        subscribedPolicies.insert(policyId)
        if connectionState == .connected {
            Task { await send(["type": "subscribe", "policyId": policyId]) }
        }
    }

    public func unsubscribe(from policyId: String) {
        subscribedPolicies.remove(policyId)
        if connectionState == .connected {
            Task { await send(["type": "unsubscribe", "policyId": policyId]) }
        }
    }

    // MARK: - Private

    private func receiveLoop() async {
        guard let task = webSocketTask else { return }
        while connectionState == .connected {
            do {
                let message = try await task.receive()
                if case .string(let text) = message, let data = text.data(using: .utf8) {
                    handleMessage(data)
                }
            } catch {
                connectionState = .disconnected
                break
            }
        }
    }

    private func handleMessage(_ data: Data) {
        guard let update = try? decoder.decode(TallyUpdate.self, from: data),
              update.type == "tally" || update.type == "snapshot" else { return }
        tallySubject.send(update)
    }

    private func send(_ dict: [String: String]) async {
        guard let task = webSocketTask,
              let data = try? JSONSerialization.data(withJSONObject: dict),
              let text = String(data: data, encoding: .utf8) else { return }
        try? await task.send(.string(text))
    }
}
